Add redux comments store

samuelclemens opened the pull request:
This PR lays some ground work for issues #2839, #967, and #792 .
Current status is that the code is mainly done, what left IMO, are some style adjustments and documentation. 
Your opinion is appreciated.

#2839:
The comments pulled by date in DESCending order, while a tree of comments is being build in the store.
This is achieved by calling `requestPostComments()` action creator, subsequent calls will fetch the earlier (older) dated comments.

#967:
The comments are built in a tree-like structure, that allows us to see in each comment which are it's children and how many children the comment has.

It can be used something like [that](https://github.com/samuelclemens/wp-calypso/blob/add/comments-redux/client/reader/comments-redux/index.jsx#L36-L50):
```
const commentChildrenIds = this.props.comments.getIn( [ this.props.commentId, 'children' ] ).toJS();

if ( commentChildrenIds.length > 0 ) {
	if ( this.state.viewReplies ) {
		return <ol>
				{ commentChildrenIds.reverse().map( ( commentId ) => <Comment ... /> ) }
				</ol>
	} else {
		return <button onClick={ this.viewRepliesClickHandler.bind( this ) }> view { commentChildrenIds.length } replies</button>;
	}
}
```

<del>
#792:
Polling can be done by invoking `pollComments()` action creator ( from `<Interval />` maybe? ) after first `requestPostComments()`, it'll fetch the new comments ( comments that got approved and meet the time-frame and entirely new comments ) and remove comments that got unapproved.

It is very inefficient however, and  it also can't detect edited comments.
</del>
No polling support in this PR.

nb commented on client/state/comments/actions.js:
```diff
@@ -0,0 +1,381 @@
+	} );
+
+	// if the api have returned comments count, dispatch it
+	if ( totalCommentsCount > -1 ) {
```
In which cases `totalCommentsCount` can be -1 (or less)?

samuelclemens replied:
Most of the times the api will return found === -1 (don't think it can be less, -1 is to indicate "unable/error" I guess.), since I use before filter on the request to fetch more comments, whenever a filter is used, api would not return the total comments count.
The only time it returns the count is when the request is first made without any filters.

nb commented on client/state/comments/test/actions.js:
```diff
@@ -0,0 +1,358 @@
+		let commentsTree;
+
+		before( () =>  {
+			const data = {"700":{"children":[739]},"701":{"children":[740]},"704":{"children":[741]},"705":{"children":[742]},"706":{"children":[743]},"707":{"children":[744]},"708":{"children":[745]},"709":{"children":[746]},"710":{"children":[747]},"711":{"children":[748]},"712":{"children":[749]},"714":{"children":[753]},"718":{"children":[755]},"728":{"children":[751]},"735":{"children":[754]},"739":{ ...
```
Making changes to this array seems very hard. Do we need all of this data? Can we make build localized, test-per-test versions?

samuelclemens replied:
Sorry I haven't wrote it earlier here, I've already addressed it a while ago by making that data [much smaller](https://github.com/Automattic/wp-calypso/pull/3143/files#diff-7683b0e3a8968ec7d9b075cd827dc5daR205).
However, I'll be looking more into how to make the tests more simpler.

nb commented on client/state/comments/test/actions.js:
```diff
@@ -0,0 +1,358 @@
+
+	describe( '#fetchAllCommentIds()', () => {
+		before( () => {
+			//nock.recorder.rec( {
```
Is there a reason to keep the commented code around?

samuelclemens replied:
I use it to record the requests, I'll remove it.

nb commented on client/state/comments/test/actions.js:
```diff
@@ -0,0 +1,358 @@
+
+	describe( '#writeComment()', () => {
+		before( () => {
+			nock('https://public-api.wordpress.com:443')
```
Why did you decide to mock the request, instead of `wpcom.js`?

samuelclemens replied:
To tell the truth, I just took a look at [posts tests](https://github.com/Automattic/wp-calypso/blob/master/client/state/posts/test/actions.js#L69) =)
It has the plus of checking if wpcom api change brakes anything for that module, imo. 

You suggesting that mock wpcom is a better approach? Could you elaborate?

nb replied:
I am mostly curious about your reasoning and the trade-offs. Both ways work.

samuelclemens replied:
@nb I think that mocking http requests has the additional benefit of testing my `actions` compliant with `wpcom` api if it might change in the future.

I've refactored the tests a bit, removed fixtures entirely and the [mock data](https://github.com/Automattic/wp-calypso/pull/3143/files#diff-7683b0e3a8968ec7d9b075cd827dc5daR202) required now is relatively small, and the mock [itself is simple](https://github.com/Automattic/wp-calypso/pull/3143/files#diff-7683b0e3a8968ec7d9b075cd827dc5daR225).

blowery commented on client/state/comments/README.md:
````diff
@@ -0,0 +1,29 @@
+
+```
+comments = {
+	items: Map<CommentTargetId, Tree>,
````
I think I'd rather nest all of the info under one Map of CommentTargetId to named bits. Having all these maps that take the CommentTargetId as the key makes it harder to just drop a set of comments and their info.

samuelclemens replied:
@blowery 
Addressed with [this](https://github.com/samuelclemens/wp-calypso/commit/866e9a3fbfd4303dddb9665410a5882d5506ed54)
Not sure if it's the solution for that.

blowery commented on client/state/comments/utils.js:
```diff
@@ -0,0 +1,58 @@
+ * @returns {String} request id
+ */
+export function createRequestId( siteId, postId, query ) {
+	return `${siteId}-${postId}-${ JSON.stringify( query ) }`;
```
I don't think we can stringify the query safely and use it as a key. key order can change which will change the key, even though the objects are equivalent.

samuelclemens replied:
You are absolutely right, 
I've addressed this [here](https://github.com/Automattic/wp-calypso/pull/3143/files#diff-b157f2e27a25c68197b5a64c67e94111R34).
+ added [a test](https://github.com/Automattic/wp-calypso/pull/3143/files#diff-a24dbf1e4112373bf10d54d63fbe1687R23) for that.

blowery commented on client/state/comments/utils.js:
```diff
@@ -0,0 +1,58 @@
+ * @param {Object} input input object
+ * @returns {Object} mapped object
+ */
+export function fieldsMapper( fields, input ) {
```
i think you can use lodash's pick instead https://lodash.com/docs#pick

samuelclemens replied:
I've removed that function and refactored any uses to `pick`.

blowery commented on client/state/comments/test/fixtures/nock-requests.json:
```diff
@@ -0,0 +1 @@
+[{"scope":"https://public-api.wordpress.com:443","method":"GET","path":"/rest/v1.1/sites/91750058/posts/287/replies/?order=DESC&number=50","body":"","status":200,"response":["1f8b0800000000000003ed9d0b93dbb6b5c7bf0abc9d9bdccc5dc900f1a49ca4935713b7899369f65e4fa6db6640125c314b912a4979adb6f9eef7007aac4451b1566b711589ad634b220812100f7e3ac0f91ffcfb22ce27597431209eb8bc2893cafcfcf2cb8b814f24c798abcb8b30 ...
```
what is this? Can we express the needs a bit more succinctly? I can't just look at this and know what's heading out.

samuelclemens replied:
Just to clarify, You are bothered with the approach of recording the API responses
or you're bothered with the response being stored as hex bytes string "1f8b0800000000000003ed9d0b93dbb6b5c7b..." ?
( that's just gzip, note the magic `0x1f8b` )

I can actually not store anything (delete that file), but that'll slow the test(s) unnecessary...

samuelclemens replied:
I've removed gzip encoding, and removed any unnecessary fields from responses.

samuelclemens replied:
@nb suggested to try get rid of all the fixtures altogether, I'll try to do that in the upcoming days.

samuelclemens replied:
@blowery 
I've refactored the tests a bit, removed fixtures entirely and the [mock data](https://github.com/Automattic/wp-calypso/pull/3143/files#diff-7683b0e3a8968ec7d9b075cd827dc5daR202) required now is relatively small, and the mock [itself is simple](https://github.com/Automattic/wp-calypso/pull/3143/files#diff-7683b0e3a8968ec7d9b075cd827dc5daR225).

blowery commented on client/state/comments/test/fixtures/mock-data.js:
```diff
@@ -0,0 +1,10 @@
+export const actualResponse = {"found":79,"site_ID":78992097,"comments":[{"ID":765,"post":{"ID":1012,"title":"The Art of Staying Still","type":"post","link":"https:\/\/public-api.wordpress.com\/rest\/v1.1\/sites\/78992097\/posts\/1012"},"author":{"ID":31299473,"login":"tessalove1","email":false,"name":"Tessa Love","first_name":"Tessa","last_name":"Love","nice_name":"tessalove1","URL":"","avatar_U ...
```
I don't think you need to have all this data in the responses. It's handy to just grab something from the API, but I would rather have specific, clean cases that express requirements in the tests.

samuelclemens replied:
That was removed, I use much simpler data now, however I'll be looking into making mock data simpler with helper functions or otherwise.

samuelclemens replied:
I've refactored the tests a bit, removed fixtures entirely and the [mock data](https://github.com/Automattic/wp-calypso/pull/3143/files#diff-7683b0e3a8968ec7d9b075cd827dc5daR202) required now is relatively small, and the mock [itself is simple](https://github.com/Automattic/wp-calypso/pull/3143/files#diff-7683b0e3a8968ec7d9b075cd827dc5daR225).

blowery commented on client/state/comments/test/actions.js:
```diff
@@ -0,0 +1,409 @@
+				expect( action.siteId ).to.equal( FEW_COMMENTS_POST.siteId );
+				expect( action.postId ).to.equal( FEW_COMMENTS_POST.postId );
+
+				done();
```
can just return the Promise instead of invoking done

samuelclemens replied:
I've [refactored](https://github.com/Automattic/wp-calypso/pull/3143/files#diff-7683b0e3a8968ec7d9b075cd827dc5daR181) the async tests to return promises.

blowery commented on client/state/comments/actions.js:
```diff
@@ -0,0 +1,381 @@
+	}
+
+	const batch = wpcom.batch();
+	const batchUrls = commentIds.map( ( commentId ) => createUrlFromTemplate( '/sites/$site/comments/$comment_ID', { site: siteId, comment_ID: commentId } ) );
```
could just use an es6 template string here rather than making a new function to do it

samuelclemens replied:
But I need the function anyway for another [case here](https://github.com/Automattic/wp-calypso/pull/3143/files#diff-a1e5174c2173e0927694ec3bd26d79d4R244).

blowery commented on client/state/comments/actions.js:
```diff
@@ -0,0 +1,381 @@
+	batchUrls.forEach( batch.add.bind( batch ) );
+
+	return batch.run().then( ( batchResultForUrls ) => {
+		let arrayOfcomments = batchUrls.map( ( batchUrl ) => batchResultForUrls[ batchUrl ] );
```
what happens if part of the batch errors?

samuelclemens replied:
@blowery I've addressed it, thanks for pointing that out.

[Here](https://github.com/Automattic/wp-calypso/pull/3143/files#diff-a1e5174c2173e0927694ec3bd26d79d4R243) and [here](https://github.com/Automattic/wp-calypso/pull/3143/files#diff-a1e5174c2173e0927694ec3bd26d79d4R282).

gwwar commented on client/state/comments/reducer.js:
```diff
@@ -0,0 +1,238 @@
+	COMMENTS_COUNT_RECEIVE,
+	COMMENTS_REQUEST,
+	COMMENTS_REQUEST_SUCCESS,
+	COMMENTS_REQUEST_FAILURE
```
It'd be a great help if SERIALIZE and DESERIALIZE is implemented in this PR. We can have the comments subtree return default initial state for both until we're ready to persist this.

samuelclemens replied:
@gwwar I've made some [initial implementation](https://github.com/Automattic/wp-calypso/pull/3143/files#diff-b3d7f2f424cf387123e9eb8f231cf85aR333) without validating schema, tell me what you think.

gwwar commented on client/state/comments/actions.js:
```diff
@@ -0,0 +1,381 @@
+ * @param {Object[]} comments comments array
+ * @param {Number} totalCommentsCount comments count that the server have on that post
+ */
+function commentsRequestSuccess( dispatch, requestId, siteId, postId, comments, totalCommentsCount ) {
```
Maybe pass though an options object instead. This signature is pretty long.

samuelclemens replied:
I've [refactored](https://github.com/Automattic/wp-calypso/pull/3143/files#diff-a1e5174c2173e0927694ec3bd26d79d4R39) it to pass an options object instead.

gwwar commented on client/state/comments/actions.js:
```diff
@@ -0,0 +1,381 @@
+		error: err
+	} );
+
+	throw err;
```
I don't think we'd want to throw here.

samuelclemens replied:
Since it's just a helper function for `requestPostComments` If I don't throw here, I silently swallow the error, it won't propagate further to whoever gets the resulting promise from `requestPostComments`.

samuelclemens replied:
On a second thought, you're right, shouldn't throw here, I already "handled" the error, no additional handling should be done anywhere else.

gwwar commented on client/state/comments/actions.js:
```diff
@@ -0,0 +1,381 @@
+}
+
+
+let placeHolderCounter = 0;
```
@blowery if we persist comments in the future we probably need to store the current counter in the tree, or use a non-incrementing id.

samuelclemens replied:
@gwwar Thanks for pointing that out, Since it's need to be unique only in the context of client, postId, siteId, [I've made it time based](https://github.com/Automattic/wp-calypso/pull/3143/files#diff-a1e5174c2173e0927694ec3bd26d79d4R144).

gwwar replied:
:+1:

gwwar commented on client/state/comments/reducer.js:
```diff
@@ -0,0 +1,266 @@
+			// the combined reducer already handled fromJS() for us, the only thing left
+			// is fix the numeric keys, which JS has as strings.
+			return state.mapKeys( ( k ) => ( ! isNaN( k ) ) ? parseInt( k, 10 ) : k );
+			break;
```
break isn't necessary here due to the return above.

samuelclemens replied:
I've removed the brake, thanks.

gwwar commented on client/state/comments/reducer.js:
```diff
@@ -0,0 +1,266 @@
+			break;
+
+		case SERIALIZE:
+			// no need to do anything here, since the combined reducer will handle it for us.
```
Maybe `return state` rather than the break? It makes it a bit clearer if new cases get added below.

samuelclemens replied:
Changed to `return`, it also inline with the other returns, thanks!

gwwar commented on client/state/comments/reducer.js:
```diff
@@ -0,0 +1,266 @@
+ * @param {Object} action redux action
+ * @returns {Object} redux state
+ */
+export function latestCommentDate( state, action ) {
```
Maybe set a default value for state here.

samuelclemens replied:
The default state is undefined until we receive appropriate data, then it becomes Date instance.
Could you give an example of what default state you had in mind?

gwwar replied:
Specifying a default value allows us to know what an unset comment date looks like. Some other subtrees choose to use `null`, `false`, `{}`, an empty Immutable.js instance etc.

samuelclemens replied:
What I meant that in the context of this specific reducer, that tracks latest comment date, `undefined` seems a good default value. since the date is not defined, until it is (when received some comments), and then we have a date object.

It doesn't makes sense to me initialize it with some value. 
I don't see any benefit on using null/false/{}/empty immutable as a default value here.

gwwar replied:
Using `undefined` works, but it unfortunately goes against Redux convention. I know you're using a different combine function, but it'd be great to follow convention when we can. See the notes section in: [combineReducers](http://redux.js.org/docs/api/combineReducers.html)

> If the state given to it is undefined, it must return the initial state for this specific reducer. According to the previous rule, the initial state must not be undefined either.

samuelclemens replied:
Ohh, I now get what you meant. That compatibility is important. I'll add a test for it. Thanks for pointing that out.
After giving it more thought, specifically the latest/earliest date reducers will be removed and replaced by selectors, the remaining ones will have initial state set to some value.

gwwar replied:
:smile: I was reading the diff linearly so it might be a moot point if we keep `combineReducersWithId`.

samuelclemens replied:
I don't think that's related =) 
It all boils down to what state structure we prefer:

```javascript
comments[ `${ site_ID }_${ post_ID }` ] = { tree, totalCommentsCount, ... }
```
^--- `combineReducersWithId` has a place

**-vs-**
```javascript
comments = { 
    tree: {
        [ `${ site_ID }_${ post_ID }` ]: { ...some comments tree... }
    },
    totalCommentsCount: {
       [ `${ site_ID }_${ post_ID }` ]: 1024
    },
    ...more props...
}
```
^--- no need for `combineReducersWithId`

samuelclemens replied:
I've moved that from reducer to a [selector](https://github.com/Automattic/wp-calypso/pull/3143/files#diff-1f1818e52092939a106eee5eaef1da11R65).

gwwar commented on client/state/comments/utils.js:
```diff
@@ -0,0 +1,127 @@
+ * @param {Function} createId id factory based on action. It should return { type: COMBINED_TARGET_TYPE, id } When the factory returns null, reducer skips processing
+ * @returns {Function} combined reducer function
+ */
+export function combineReducersWithId( reducers, createId ) {
```
Could you provide a few examples of what problems this new combine reducers function solves?

samuelclemens replied:
@gwwar [Sure!](https://github.com/Automattic/wp-calypso/pull/3143/files#diff-b157f2e27a25c68197b5a64c67e94111R68)

What I'm trying to solve here is to create a state tree of the following form:
```javascript
comments[ `${ site_ID }_${ post_ID }` ] = { tree, totalCommentsCount, ... }
```
instead of:
```javascript
comments = { 
	tree: {
		[ `${ site_ID }_${ post_ID }` ]: { ...some comments tree... }
	},
	totalCommentsCount: {
	   [ `${ site_ID }_${ post_ID }` ]: 1024
	},
	...more props...
}
```

Also discussed [here](https://github.com/Automattic/wp-calypso/pull/3143#discussion_r52378821).

gwwar replied:
Interesting, but I would still prefer us to use the standard reducer. Constructing the second data structure should be possible, though it requires us to add handlers for `COMMENTS_RECEIVE` and `COMMENTS_REMOVE` on the reducer for all comments subtrees.

@aduth any feelings on this?

samuelclemens replied:
I'm not really sure what you mean, I'll try explain more where I'm coming from.

Let's assume that we want the following structure for comments state:
```javascript
comments[ `${ site_ID }_${ post_ID }` ] = { tree, totalCommentsCount, ... }
```
Where `comments` is `Immutable.Map`.

A reducer for that state structure will look something like that (psuedo-code):
```javascript
// comments.js
export default function reducer( state = Immutable.Map(), action ) {
	if ( ! action.siteId || ! action.postId ) {
		return state;
	}	
	const id = action.siteId + '_' + action.postId;

	// ensure we have a map for the id
	let newState = state;
	if ( ! newState.get( id ) ) {
		newState = newState.set( id, Immutable.Map() );
	}
	
	switch ( action.type ) {
			case COMMENTS_RECEIVE:
				return newState.setIn( [ id, 'tree' ], tree( action ) );
			case COMMENTS_COUNT_RECEIVE:
				return newState.setIn( [ id, 'totalCommentsCount' ], totalCommentsCount( action ) );
			
			... more action(s)...
	}
	
	// nothing changed, return original
	return state;
}
```

Notice the `tree` and `totalCommentsCount` functions, they will actually perform the state change,
as we will have more actions, we'll have to add them to that reducer, as a `CASE`, and create
a function that will handle it. Other option ofcurse is have the functions inline, in `totalCommentsCount` case
that doesn't really matter as it's really tiny, but at tree's case it does.
IMO, it also creates a long reducer function that's harder to follow, and I prefer to avoid that if possible.

`combineReducersWithId()` is simply a general version of the above reducer, it has nothing special about it,
it just creates a `comments( state, action ) = newState` function.

You've mentioned you 
> prefer to use the standard reducer

I'm not really sure what you mean, did you mean you'd like to use a _long_ reducer function as an the example above?

Or just use many shorter reducer(s) that create the form of:
```javascript
comments = Map { 
    tree: Map {
        [ `${ site_ID }_${ post_ID }` ]: { ...some comments tree... }
    },
    totalCommentsCount: Map {
       [ `${ site_ID }_${ post_ID }` ]: 1024
    },
    ...more props...
}
```

Or another option?

Thank you for your time reviewing this. =)

gwwar replied:
Unless we have very unique concerns, I was suggesting that we use the second data structure, and stick with Redux's [combineReducers](http://redux.js.org/docs/api/combineReducers.html) function. I'm mostly hesitant because it adds a bit of developer context switching when interacting with this subtree. Let's wait for a few more folks to chime in, before you refactor anything though. :smile: 

```
comments = { 
    tree: {
        [ `${ site_ID }_${ post_ID }` ]: { ...some comments tree... }
    },
    totalCommentsCount: {
       [ `${ site_ID }_${ post_ID }` ]: 1024
    },
    ...more props...
}
```

Also, you may find the follow section useful: https://github.com/Automattic/wp-calypso/blob/master/docs/our-approach-to-data.md#data-normalization

samuelclemens replied:
@aduth @blowery any feedback on that before I'm refactoring it to @gwwar 's suggestion?

gwwar replied:
For anyone :eyes: The TLDR; is basically should we let subtrees use custom combineReducer functions? In terms of Redux, we can swap combineReducers with whatever we like, but using different types of combineReducers together will incur some developer context switching costs.

blowery replied:
If that's what we're doing elsewhere, I'm fine with it. I'd rather be consistent.

gwwar replied:
I believe that this is the first instance we have of it, so it's worth a discussion.

samuelclemens replied:
I've refactored it to the rest of calypso's convention. 
I think it might be a good convention discussion in a broader forum/dedicated issue, but for now, for the comments it might be best for me to stick to the current convention.

gwwar commented on client/state/comments/actions.js:
```diff
@@ -0,0 +1,323 @@
+	COMMENTS_REQUEST,
+	COMMENTS_REQUEST_SUCCESS,
+	COMMENTS_REQUEST_FAILURE,
+	COMMENTS_REMOVE_COMMENT,
```
Our action naming guidelines suggest that the verb is suffixed. https://github.com/Automattic/wp-calypso/blob/master/docs/our-approach-to-data.md#actions

So this can be shortened to `COMMENTS_REMOVE`

samuelclemens replied:
Changed that, thanks.

blowery commented:
We just landed a move to lodash 4.x, so we'll need to rework this to take that into account. See https://github.com/lodash/lodash/wiki/Changelog#compatibility-warnings for a list of changes.

samuelclemens commented:
I've removed polling code from this PR. 

@blowery suggested that polling all comment IDs would rise a performance problem, without that no other way to really poll, unless we change the API, which I understood is not currently an option.

Refactored the state tree to the structure @gwwar suggested, as per 
https://github.com/Automattic/wp-calypso/blob/master/docs/our-approach-to-data.md#data-normalization

Rebased.

gwwar commented on CREDITS.md:
````diff
@@ -1330,3 +1330,29 @@ LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 ```
+
+### https://github.com/reactjs/reselect
+#### client/state/comments
````
Since usages change over time, please remove ```#### client/state/comments```

samuelclemens replied:
I'll remove that, I just saw it in other libs there...

gwwar commented on client/state/comments/reducer.js:
```diff
@@ -0,0 +1,176 @@
+ */
+function getLikeSetter( { like = true, likeCount } ) {
+	return ( comment ) => comment.set( 'i_like', like )
+								.update( 'like_count', ( ct ) => {
```
Perhaps `count` over `ct`?

samuelclemens replied:
Sure, I'll change that.

gwwar commented on client/state/comments/utils.js:
```diff
@@ -0,0 +1,114 @@
+ * @param {Number} postId post identification
+ * @returns {String} comment target id
+ */
+export function createCommentTargetId( siteId, postId ) {
```
Maybe this is a bit clearer to call this a `getCommentParentKey`?

samuelclemens replied:
Sure, I'll change that.

gwwar commented on client/state/comments/reducer.js:
```diff
@@ -0,0 +1,176 @@
+ * @param {Function|Object} updaterOrValue an updater function or a value
+ * @returns {Immutable.Map} new state
+ */
+function updateSpecificState( state, action, updaterOrValue ) {
```
How about `updateCommentState( state, id, updaterOrValue )`? Since we're only using `action` to generate an id, it makes it clearer if we change the function signature to accept the id instead. It should make mocking the tests a bit shorter too.

samuelclemens replied:
The reason I've done it like that is because I didn't want to call `createCommentTargetId( action.siteId, action.postId )` in every reducer and capture the logic of getting the specific state (and the generation of it's ID) in a single place, It is after all a helper function =)

gwwar replied:
If every handler uses the id, you can create the id before the switch statement.

samuelclemens replied:
@gwwar  Every reducer needs the id, not just every handler at the switch...

samuelclemens replied:
I'll elaborate a bit more.

If I understand you correctly, you're suggesting factoring out the id generation from the `updateSpecificState`  function,

What it'll mean in my opinion, that in each reducer ( `items`, `requests`, `totalCommentsCount` ),
i'll have to repeat ID creation:
 ```
const id = createCommentTargetId( action.siteId, action.postId `); // DRY
switch ( action.type ) {
...
}
```
in my opinion, it's better not to repeat.

gwwar replied:
Passing exactly what the function needs, helps keep the intent clearer and makes refactoring easier. This is a helper function, so not a blocker here.

gwwar commented on client/state/comments/reducer.js:
```diff
@@ -0,0 +1,176 @@
+			const newIds = Immutable.Set( action.comments.map( comment => comment.ID ) );
+
+			// we prefer freshly retrieved data, so throw away old data
+			return updateSpecificState( state, action, ( oldSpecificState = Immutable.List() ) =>
```
How about `comments` instead of `oldSpecificState`?

samuelclemens replied:
Sure, I'll change that.

gwwar commented on client/state/comments/reducer.js:
```diff
@@ -0,0 +1,176 @@
+			);
+		case COMMENTS_LIKE:
+			return updateSpecificState( state, action, ( oldSpecificState = Immutable.List() ) =>
+				listFindAndModify( oldSpecificState,
```
Is there any reason why we don't use a filter here, and then modify the filtered result?

gwwar replied:
Since we're modifying a single item, it may be more appropriate to use setIn or mergeIn here.

samuelclemens replied:
I can't simply use a `sentIn` because I need to know the index on which to set (the correct comment on which to update the like), to know the index i need to find it, after I find the index I `setIn`.

That logic is encapsulated in `listFindAndModify`.

I might be missing something here since I do expect Immutable.js to handle such case, so maybe you can give an example on how to achieve that?

I'm not really following your filtering suggestion, a `filter()` creates a new collection, any modifications on that collection (or it's descendant) should not modify the old collection, that's the Immutable thing, so I can't see how that can work, again, I might be missing something here so lets chat about that =)

gwwar commented:
I'm still :eyes: I'll try to finish up tomorrow.

samuelclemens commented:
@gwwar Thanks for reviewing that!

blowery commented on client/state/comments/utils.js:
```diff
@@ -0,0 +1,114 @@
+ * @param {Boolean} updateAll whether update all matches of predicate or just the first
+ * @returns {Immutable.List} modified list
+ */
+export function listFindAndModify( list, predicate, updater, updateAll = false ) {
```
You could probably do the single item case with `findEntry` and `mergeIn`. Where do you use the multi item case?

samuelclemens replied:
I don't use the multi case.
It's still 2 not chain-able method calls I need to encapsulate in some function.

You're suggesting remove the multi case?

blowery replied:
> You're suggesting remove the multi case?

If we're not using it, yeah. No reason to have it.

> It's still 2 not chain-able method calls I need to encapsulate in some function.

Having the function is fine. Just thinking we can simplify it and use more idiomatic immutable methods.

Maybe for a name `updateExistingIn`? Given how Immutable works, this could really work for List or Map.. It's pretty much like `updateIn`, but it wouldn't create missing keys.

samuelclemens replied:
Sure, I'll make that later tonight.

gwwar commented on client/state/comments/reducer.js:
```diff
@@ -0,0 +1,176 @@
+		case SERIALIZE:
+			return state.toJS();
+		case DESERIALIZE:
+			return Immutable.fromJS( state );
```
If we want to persist this state tree, we'll want to add a JSON Schema for each subtree. We can also defer this work by temporarily choosing to not persist state.

```javascript
case SERIALIZE:
    return {};
case DESERIALIZE:
    return Immutable.Map();
```
The documentation update is still open, but if you're interested you can read some of the reasoning here: https://github.com/Automattic/wp-calypso/pull/3267

blowery replied:
I'd hold off on persisting this bit of the state tree for now.

samuelclemens replied:
I've removed the serialization.

gwwar commented on client/state/comments/actions.js:
```diff
@@ -0,0 +1,303 @@
+			commentId,
+			iLike: data.i_like,
+			likeCount: data.like_count
+		} ) ).catch( () => dispatch( {
```
:+1: I like the handling here. Do we run into any issues though if  we get an error before the `like()`? For example, we fail to fetch ```wpcom.site( siteId ).comment( commentId )```

blowery replied:
I don't think that actually fetches, it just returns an object that can be used to chain further API things. If it fetched, it would have to take a callback or hand back a Promise, which it doesn't appear to do.

We may need to validate that an successful like falls through into catch though. Separately, try liking a comment that  doesn't exist to see how it responds.

samuelclemens replied:
@gwwar That doesn't fetches, it can be seen in the tests, since `nock` blocks all outgoing requests.

@blowery I'm not sure I'm following. In order for a user to be able to 'like' a comment they need to click on the like button, which rendered on a comment, therefor the comment is in state. 

For the sake of argument let's say he/she somehow succeed in doing so, were we expect to see a result of such like? Since we don't have a comment for it, the current reducer will just ignore it.

I also not sure I'm following your comment regarding the `catch`, currently the way it works is:
- User clicks the like button, that dispatches immidiate +1 on the like count and i_like -> true. 
- Somewhat later, we get a success result from wpcom, that result contains updated likes count (including the like above and maybe some more likes), we set that count for the comment.
- We might also get a failure from wpcom, then we need to remove the +1 we made and i_like -> false.

at least imo, that's what the code does =)

blowery replied:
My comment was more on the semantics of the Promise. I did some more poking and this is fine.

gwwar commented on client/state/comments/selectors.js:
```diff
@@ -0,0 +1,118 @@
+/***
+ * External dependencies
+ */
+import { createSelector } from 'reselect';
```
:sparkles: Also, https://github.com/Automattic/wp-calypso/pull/3522 is open, so whichever PR lands first will probably need to rebase.

gwwar commented on client/state/comments/selectors.js:
```diff
@@ -0,0 +1,118 @@
+ */
+export const getCommentLike = createSelector(
+	getPostCommentItems,
+	( state, props ) => props.commentId,
```
I think this OK, but somewhat frowned upon: https://github.com/reactjs/reselect#q-how-do-i-create-a-selector-that-takes-an-argument

samuelclemens replied:
@gwwar Not really sure what you mean.
That's taken from [their example](https://github.com/reactjs/reselect#accessing-react-props-in-selectors). 
```javascript
const maxTodosSelector = (state, props) => props.maxTodos
const visibleTodosSelector = createSelector(
  ...
  maxTodosSelector,
  (visibilityFilter, todos, maxTodos) => {
   ...
  }
)
```

gwwar commented on client/state/comments/test/reducer.js:
```diff
@@ -0,0 +1,280 @@
+				type: COMMENTS_RECEIVE,
+				siteId: 1,
+				postId: 1,
+				comments: [ ...commentsNestedTree ].sort( () => Math.random() * 2 % 2 ? -1 : 1 )
```
:game_die:

gwwar commented on client/state/comments/utils.js:
```diff
@@ -0,0 +1,114 @@
+ * @returns {Immutable.Map} Immutable map comments tree instance of the shape Map<id, CommentNode>{ children: List<id> }
+ */
+export function buildCommentsTree( comments ) {
+	const aTree = Immutable.fromJS( { children: [] } );
```
Tiny nit, but maybe `tree` or `commentsTree` instead of `aTree`.

samuelclemens replied:
I changed that, thanks.

gwwar commented:
:+1: Great progress! Before merging
- [ ] Wait on #3545 or #3522 to land. There may be a little bit of refactoring if #3545 lands.
- [ ] Don't persist the subtree, or add schemas https://github.com/Automattic/wp-calypso/blob/master/docs/our-approach-to-data.md#data-persistence--2754-

blowery commented:
Looks like #3545 landed, so we should switch over to that from `reselect`.

samuelclemens commented:
Migrated to #3545, blocked by #3627 or alternative.

samuelclemens commented:
#3627 is merged, so this PR should be good to go imo.

blowery commented:
Looks good to me. @gwwar any objections to landing this?

gwwar commented:
:+1: :ship:

blowery commented:
:ship: :tada: :balloon:
